cli/Makefile

```diff
@@ -54,13 +53,34 @@ $(BUILDDIR)/julia_res.o: $(JULIAHOME)/contrib/windows/julia.rc $(JULIAHOME)/VERS
 		'/^(\d+)\.?(\d*)\.?(\d*)/ && \
 		print int $$1,",",int $$2,",0,",int $$3'` && \
 	$(CROSS_COMPILE)windres $< -O coff -o $@ -DJLVER=$$JLVERi -DJLVER_STR=\\\"$$JLVER\\\"
-OBJS += julia_res.o
-DOBJS += julia_res.o
+OBJS += $(BUILDDIR)/julia_res.o
+DOBJS += $(BUILDDIR)/julia_res.o
 endif
 endif
 
-julia-release: $(build_bindir)/julia$(EXE) $(build_shlibdir)/libjulialoader.$(JL_MAJOR_MINOR_SHLIB_EXT)
-julia-debug: $(build_bindir)/julia-debug$(EXE) $(build_shlibdir)/libjulialoader-debug.$(JL_MAJOR_MINOR_SHLIB_EXT)
+ifneq ($(OS),WINNT)
+HIDDEN_SYMBOL := __attribute__((visibility(\"hidden\")))
+endif
+
+# Create a list of all `jl_` symbols that libjulia-internals contains that libjulia will re-export
+$(BUILDDIR)/jl_exports.h: $(SRCDIR)/../src/jl_exported_funcs.list $(SRCDIR)/../src/jl_exported_data.list
+	@echo "// This list of symbols auto-generated by cli/Makefile and sourced from src/jl_exported_*.list; do not edit" > $@
+	@echo "// Define Data symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_data.list | awk '{ print "JL_DLLEXPORT void * " $$1 ";" }' >> $@
+	@echo "// Function symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "$(HIDDEN_SYMBOL) void * " $$1 "_addr;" }' >> $@
+	@echo >> $@
+	@echo "void loader_export_symbols(void * libjulia_internal) {" >> $@
+	@echo "    // Import function symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "    " $$1 "_addr = lookup_symbol(libjulia_internal, \"" $$1 "\");" }' >> $@
+	@echo "}" >> $@
```

Comment:
To avoid the need for awk as a preprocessor, it's perhaps a bit more typically to use the c preprocessor (example adapted from uv.h. You can find similar in LLVM, though often with #include instead.):
```c
#define JL_FUNCTION_MAP(XX) \
  XX(jl_) \
  XX(jl_abs_float) \
  XX(jl_abs_float_withtype) \
  XX(jl_add_float) \
  XX(jl_add_int) \
  XX(jl_add_optimization_passes) \
...

#define STR(a) #a
#define XSTR(x) STR(x)

#if defined _CPU_X86_64_
#define JL_DLLREEXPORT(f) \
    JL_DLLEXPORT JL_NAKED void f(void) { __asm__(\"jmpq *%0\" : : \"a\"f#_addr : ) }
#elif defined _CPU_X86_
#define JL_DLLREEXPORT(f) \
    JL_DLLEXPORT JL_NAKED void f(void) { __asm__(\"jmpl *%0\" : : \"a\"f#_addr : ) }
#else
#warning "trampoline support needs to be defined for this CPU to enable linking against libjulia"
#undef JL_FUNCTION_MAP
#define JL_FUNCTION_MAP(x)
#endif

JL_FUNCTION_MAP(JL_DLLREEXPORT)
#undef JL_DLLREEXPORT

#define XX(f) static void *#f#_addr;
JL_FUNCTION_MAP(XX)
#undef XX

#define XX(f) XSTR(f),
static const char *const func_names[] = {
JL_FUNCTION_MAP(XX)
NULL
};
#undef XX

#define XX(f) &#f#_addr,
static void *const *const func_addrs[] = {
JL_FUNCTION_MAP(XX)
NULL
};
#undef XX

void loader_export_symbols(void * libjulia_internal)
{
    auto names = func_names;
    auto addrs = func_addrs;
    while (*names) {
        *addrs++ = dlsym(libjulia_internal, *names++);
    }
}
```